### ISSUE TITLE
Provide ability to "always" autoformat part of a file?

### ISSUE THREAD

**Issue by andrewferrier** [Chiel92/vim-autoformat#132]
I have two bindings for `vim-autoformat` in my `.vimrc` like this:

nnoremap <Leader>f :Autoformat<CR>
vnoremap <Leader>f :Autoformat<CR>

I was hoping that the visual-mode binding would reformat just the lines/area selected, but it doesn't seem to do that by default (for example, for `.html`).

Upon reading the "Ranged definitions" section of README.md, I think I understand why; the default formatdefs for htmlbeautify and tidy don't support `a:firstline` or `a:lastline` - presumably because tidy/htmlbeautify themselves don't support this.

This made me think - a useful default behaviour for `vim-autoformat` in this case would be to fallback to piping only the visually highlighted portion of the file to those formatters instead. In the case of HTML, for example, a small section of the document, if correctly selected, is often syntactically valid by itself, and so can be formatted in isolation.

What do you think?

**Comment by Chiel92** [comment]
Yeah, some formatters could deal with that I think. Just curious, is there a specific reason you don't want to format the rest of the file?

**Comment by andrewferrier** [comment]
@Chiel92 mostly it's generally because I'm dealing with some existing source code, often badly formatted or laid out. I don't want to reformat the whole file as it will completely break source code diffing if it's checked in like that; I just want to reformat the part I'm working on/changing.

**Comment by Chiel92** [comment]
The problem we're facing here is that we have to make clear which formatters need the entire file when in visual mode, and which formatters only need the selected text.

**Comment by andrewferrier** [comment]
Perhaps there could be an option to turn this on formatter by formatter? Or alternatively just "assume" it will be OK, and treat cases where the formatter fails on the text "subset" like any other failure.

**Comment by Chiel92** [comment]
>  Or alternatively just "assume" it will be OK, and treat cases where the formatter fails on the text "subset" like any other failure.

How would you know when a formatter fails? It may just silently format the code bogusly.

Another hacky way would be to check if the range parameters are used and if not, only pass the subset.

**Comment by andrewferrier** [comment]
I was assuming the user would notice the bogus result :)

**Comment by Chiel92** [comment]
Ah, I thought you meant automatic recognition by vim-autoformat, which would then fallback to passing the entire text (with the range parameters).

I'm gonna think about it for a bit, and probably will try and see whether my second suggestion works. That one would be the least intrusive.

**Comment by fwip** [comment]
Another suggestion: Pass the entire file to the formatter, then try to apply only the changes in the selected region to the file.

This may be non-trivial or error-prone if the file is badly formatted as a whole, but might work alright if the file is generally properly formatted.

**Comment by pseewald** [comment]
I'd advice against passing the entire file to formatters since many formatters split lines or remove empty lines, then it's not easily possible to recover the section of interest after formatting.

I agree with @andrewferrier that being able to selectively format is an essential feature. I would not worry whether formatters are able to format selections of code, I think it's better if all formatters behave identically. If it fails, it is the responsibility of the user to not apply a formatter to a visual selection.

**Comment by Rikorose** [comment]
Another possibility would be to take the lines to format as a list. This would be really helpful to only format lines that have been changed. One could get the changed lines from some git plugin.

For instance `clang-format` takes the `lines` argument:
```
  -lines=<string>           - <start line>:<end line> - format a range of
                              lines (both 1-based).
                              Multiple ranges can be formatted by specifying
                              several -lines arguments.
                              Can't be used with -offset and -length.
                              Can only be used with one input file.

```

**Comment by MaiLunJiye** [comment]
> Yeah, some formatters could deal with that I think. Just curious, is there a specific reason you don't want to format the rest of the file?

markdown with many different languages code block,and i want to use different formatter to format the code blocks like:

    this is markdown

    ```c
    int main(){
        print("please use clang-format to format me");
        return 0;
    }
    ```

    some python

    ```python
    print("please use autopep8 to format me")
    ```

i hope we can select the block in visual model, and then use formatter what we give(args) to format what we select

**Comment by Chiel92** [comment]
Yes, we could do this. But we would need to make a distinction between "ranged definitions" and "normal definitions" to make this work correctly and fluently.

Given a formatter, if it is a ranged definition, we can simply proceed as we do now. If it is not a ranged formatter, and we do have a visual selection, then only format and replace the selected lines of the buffer.

This is a breaking change unfortunately, and it needs a bit of work.
